46.md

```diff
@@ -2,161 +2,67 @@ NIP-46
+`draft` `optional` `author:tiero` `author:giowe` `author:vforvalerio87` `author:pablof7z` `author:fiatjaf`
 
-## Rationale
+This NIP describes a method for 2-way communication between a **remote signer** and a normal Nostr client. The remote signer could be, for example, a hardware device dedicated to signing Nostr events, while the client is a normal Nostr client.
 
-Private keys should be exposed to as few systems - apps, operating systems, devices - as possible as each system adds to the attack surface.
+The remote signer generates a connection token in the form
 
-Entering private keys can also be annoying and requires exposing them to even more systems such as the operating system's clipboard that might be monitored by malicious apps.
-
-
-## Terms
-
-* **App**: Nostr app on any platform that *requires* to act on behalf of a nostr account.
-* **Signer**: Nostr app that holds the private key of a nostr account and *can sign* on its behalf.
-
-
-## `TL;DR`
-
-
-**App** and **Signer** sends ephemeral encrypted messages to each other using kind `24133`, using a relay of choice. 
-
-App prompts the Signer to do things such as fetching the public key or signing events.
+```
+<npub1...>#<secret>?relay=wss://...&relay=wss://...
+```
 
-The `content` field must be an encrypted JSONRPC-ish **request** or **response**.
+The user copies that token and pastes it in the client UI somehow. Then the client can send events of kind `24133` to the specified relays and wait for responses from the remote signer.
 
-## Signer Protocol
+The `secret` allows the signer to identify the client's key and, perhaps, assign it certain permissions (e.g. *sign only event `kind:7`, reject encrypt/decrypt requests*)
 
-### Messages
+Event payloads are [NIP-04](04.md)-encrypted JSON blobs that look like JSONRPC.
 
-#### Request
+Events sent by the client to the remote signer have the following format:
 
-```json
+```js
 {
-  "id": <random_string>,
-  "method": <one_of_the_methods>,
-  "params": [<anything>, <else>]
+  "pubkey": "<client-deterministically-generated-key-for-the-connection>"
+  "kind": 24133,
+  "tags": [
+    ["p", "<user-pubkey-hex>"]
+  ],
+  "content": "nip04_encrypted_json({id: <random-string>, method: <see-below>, params: []})",
+  ...
 }
 ```
 
-#### Response
+And the events the remote signer sends to the client have the following format:
 
-```json
-{
-  "id": <request_id>,
-  "result": <anything>,
-  "error": <reason>
-}
+```js
+  "pubkey": "<user-pubkey>"
+  "kind": 24133,
+  "tags": [
+    ["p", "<signer-pubkey-hex>"]
+  ],
+  "content": "nip04_encrypted_json({id: <request-id>, result: <any>, error: <reason-string>})",
+  ...
 ```
 
 ### Methods
 
-
-#### Mandatory 
-
-These are mandatory methods the remote signer app MUST implement:
-
-- **describe**
-  - params []
-  - result `["describe", "get_public_key", "sign_event", "connect", "disconnect", "delegate", ...]`  
+- **connect**
```

Comment:
Is the connect event still needed given the user now pastes the key from the signer into the app?

Reply:
I agree, but I think implementations out there are relying on this? @pablof7z @staab

Reply:
I don't know, I use pablo's implementation

Reply:
if the user is not sending a secret to apply a pre-authorized list the `connect` method is redundant; nsecBunker doesn't depend on it. Ok with me if we make it optional

Reply:
Can we just remove it?

Reply:
I think connect is nice to have for future connections, just so client can verify that it can reach the signer (and if not notify the user). however I am not opposed to removing it entirely. 

If it is not removed, we should at least remove the params (since the signer should be able to identify the client from the event's pubkey already)